src/driver/amdxdna/aie2_pci.c

```diff
@@ -54,11 +78,56 @@ static inline void aie2_dump_chann_info_debug(struct amdxdna_dev_hdl *ndev)
 	XDNA_DBG(xdna, "x2i ringbuf 0x%x", ndev->mgmt_x2i.rb_start_addr);
 	XDNA_DBG(xdna, "x2i rsize   0x%x", ndev->mgmt_x2i.rb_size);
 	XDNA_DBG(xdna, "x2i chann index 0x%x", ndev->mgmt_chan_idx);
+	if (!ndev->mgmt_prot_major)
+		return;
+
+	XDNA_DBG(xdna, "mailbox protocol major 0x%x", ndev->mgmt_prot_major);
+	XDNA_DBG(xdna, "mailbox protocol minor 0x%x", ndev->mgmt_prot_minor);
+}
+
+static bool aie2_is_legacy_mgmt_info(struct amdxdna_dev_hdl *ndev, u32 off)
+{
+	u32 *mgmt_info_bar_addr = ndev->sram_base + off;
+	u32 magic;
+
+	magic = readl(mgmt_info_bar_addr + MAGIC_OFFSET);
+	if (strcmp((const char *)&magic, MGMT_MBOX_MAGIC))
+		return true;
+
+	return false; /* Magic number found */
+}
+
+static void aie2_get_mgmt_chann_info_legacy(struct amdxdna_dev_hdl *ndev, u32 off)
+{
+	struct mgmt_mbox_chann_info_lagcy info_regs;
+	struct xdna_mailbox_chann_res *i2x;
+	struct xdna_mailbox_chann_res *x2i;
+	u32 *reg;
+	int i;
+
+	reg = (u32 *)&info_regs;
+	for (i = 0; i < sizeof(info_regs) / sizeof(u32); i++)
+		reg[i] = readl(ndev->sram_base + off + i * sizeof(u32));
+
+	i2x = &ndev->mgmt_i2x;
+	x2i = &ndev->mgmt_x2i;
+
+	i2x->mb_head_ptr_reg = AIE2_MBOX_OFF(ndev, info_regs.i2x_head);
+	i2x->mb_tail_ptr_reg = AIE2_MBOX_OFF(ndev, info_regs.i2x_tail);
+	i2x->rb_start_addr   = AIE2_SRAM_OFF(ndev, info_regs.i2x_buf);
+	i2x->rb_size         = info_regs.i2x_buf_sz;
+
+	x2i->mb_head_ptr_reg = AIE2_MBOX_OFF(ndev, info_regs.x2i_head);
+	x2i->mb_tail_ptr_reg = AIE2_MBOX_OFF(ndev, info_regs.x2i_tail);
+	x2i->rb_start_addr   = AIE2_SRAM_OFF(ndev, info_regs.x2i_buf);
+	x2i->rb_size         = info_regs.x2i_buf_sz;
```

Comment:
pls do not duplicate these lines of code. just fill them in as before and when you reach the end, read the magic u32 to decide if you need to read more or not. all of these can be done in one function.

Reply:
Done.